pkg/disttask/framework/metering/metering.go

```diff
@@ -146,25 +101,105 @@ func NewMeter(cfg *mconfig.MeteringConfig) (*Meter, error) {
 	}
 	meteringConfig := mconfig.DefaultConfig().WithLogger(logger)
 	writer := meteringwriter.NewMeteringWriterFromConfig(provider, meteringConfig, cfg)
+	return newMeterWithWriter(logger, writer), nil
+}
+
+func newMeterWithWriter(logger *zap.Logger, writer meteringwriterapi.MeteringWriter) *Meter {
 	return &Meter{
-		logger: logger,
-		data:   make(map[int64]*Data),
-		writer: writer,
-		uuid:   strings.ReplaceAll(uuid.New().String(), "-", "_"), // no dash in the metering sdk
-	}, nil
+		logger:          logger,
+		recorders:       make(map[int64]*wrappedRecorder),
+		lastFlushedData: make(map[int64]*Data),
+		writer:          writer,
+		uuid:            strings.ReplaceAll(uuid.New().String(), "-", "_"), // no dash in the metering sdk
+	}
 }
 
-// Record records a meter data.
-func (m *Meter) Record(keyspace string, taskType TaskType, taskID int64, other *Data) {
+func (m *Meter) getOrRegisterRecorder(r *Recorder) *Recorder {
 	m.Lock()
 	defer m.Unlock()
-	if m.data[taskID] == nil {
-		m.data[taskID] = &Data{
-			keyspace: keyspace,
-			taskType: taskType,
+	if old, ok := m.recorders[r.taskID]; ok {
+		// each task might have different steps, it's possible for below sequence
+		//  - step 1 get recorder
+		//  - step 1 executor exist, and unregisters recorder, but not flushed yet
+		//  - step 2 get recorder again, we should reset the unregistered flag
+		if old.unregistered {
+			old.unregistered = false
 		}
+		return old.Recorder
+	}
+	m.recorders[r.taskID] = &wrappedRecorder{Recorder: r}
+	return r
+}
+
+// UnregisterRecorder unregisters a recorder.
+func (m *Meter) unregisterRecorder(taskID int64) {
+	m.Lock()
+	defer m.Unlock()
+	// we still need to flush for the unregistered recorder once more, so we only
+	// mark it here, and delete when it's flushed.
+	if r, ok := m.recorders[taskID]; ok {
+		r.unregistered = true
 	}
-	m.data[taskID].merge(other)
+}
+
+func (m *Meter) cleanupUnregisteredRecorders() []*Recorder {
+	removed := make([]*Recorder, 0, 1)
+	m.Lock()
+	defer m.Unlock()
+	for taskID, r := range m.recorders {
+		if !r.unregistered {
+			continue
+		}
+		// since register and flush run in async, it's possible that:
+		//  - flush start, and scrape current data(without recorder R)
+		//  - register recorder R, and unregister fast
+		//  - flush finish, so here lastFlushedData doesn't contain R, we should
+		//    keep the recorder and do a final flush.
+		if fd, ok := m.lastFlushedData[taskID]; ok {
+			// unregister and scrape is run in async, it's possible there are still
+			// some non-flushed data even the recorder is unregistered, so we check
+			// current data too.
+			if fd.equals(r.currData()) {
+				delete(m.recorders, r.taskID)
+				removed = append(removed, r.Recorder)
+			}
+		}
+	}
+
+	return removed
+}
+
+func (m *Meter) onSuccessFlush(flushedData map[int64]*Data) {
+	m.lastFlushedData = flushedData
```

Comment:
Should we lock for lastFlushedData? Will it race with L158?

Reply:
`cleanupUnregisteredRecorders` is only called within the same method